File: devtools/cliphold.c
// cliphold — nested-gate clipboard source that deliberately delays transfer.
// It uses wlr-data-control so no surface or keyboard focus is needed.
//
//   cliphold DELAY_MS TEXT
//
// A negative delay holds the requested fd until the process or compositor
// exits. READY is printed after the selection belongs to this client; SEND is
// printed when a consumer requests it.
#define _GNU_SOURCE
#include <wayland-client.h>
#include "cliphold-proto.h"

#include <errno.h>
#include <poll.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

static struct zwlr_data_control_manager_v1* manager;
static struct wl_seat*                      seat;
static int                                  send_fd = -1;
static int                                  delay_ms;
static int64_t                              send_at;
static const char*                          payload;
static int                                  cancelled;

static int64_t monotonicMs(void) {
    struct timespec t;
    clock_gettime(CLOCK_MONOTONIC, &t);
    return (int64_t)t.tv_sec * 1000 + t.tv_nsec / 1000000;
}

static void registryAdd(void* data, struct wl_registry* registry, uint32_t name, const char* interface, uint32_t version) {
    (void)data;
    if (!strcmp(interface, zwlr_data_control_manager_v1_interface.name))
        manager = wl_registry_bind(registry, name, &zwlr_data_control_manager_v1_interface, version < 2 ? version : 2);
    else if (!strcmp(interface, wl_seat_interface.name))
        seat = wl_registry_bind(registry, name, &wl_seat_interface, version < 7 ? version : 7);
}

static void registryRemove(void* data, struct wl_registry* registry, uint32_t name) {
    (void)data;
    (void)registry;
    (void)name;
}

static const struct wl_registry_listener registryListener = {registryAdd, registryRemove};

static void sourceSend(void* data, struct zwlr_data_control_source_v1* source, const char* mime, int32_t fd) {
    (void)data;
    (void)source;
    (void)mime;
    if (send_fd >= 0)
        close(send_fd);
    send_fd = fd;
    send_at = monotonicMs() + (delay_ms < 0 ? 0 : delay_ms);
    puts("SEND");
    fflush(stdout);
}

static void sourceCancelled(void* data, struct zwlr_data_control_source_v1* source) {
    (void)data;
    (void)source;
    cancelled = 1;
}

static const struct zwlr_data_control_source_v1_listener sourceListener = {sourceSend, sourceCancelled};

static void offerMime(void* data, struct zwlr_data_control_offer_v1* offer, const char* mime) {
    (void)data;
    (void)offer;
    (void)mime;
}

static const struct zwlr_data_control_offer_v1_listener offerListener = {offerMime};

static void deviceOffer(void* data, struct zwlr_data_control_device_v1* device, struct zwlr_data_control_offer_v1* offer) {
    (void)data;
    (void)device;
    zwlr_data_control_offer_v1_add_listener(offer, &offerListener, NULL);
}

static void deviceSelection(void* data, struct zwlr_data_control_device_v1* device, struct zwlr_data_control_offer_v1* offer) {
    (void)data;
    (void)device;
    if (offer)
        zwlr_data_control_offer_v1_destroy(offer);
}

static void deviceFinished(void* data, struct zwlr_data_control_device_v1* device) {
    (void)data;
    (void)device;
    cancelled = 1;
}

static void devicePrimary(void* data, struct zwlr_data_control_device_v1* device, struct zwlr_data_control_offer_v1* offer) {
    deviceSelection(data, device, offer);
}

static const struct zwlr_data_control_device_v1_listener deviceListener = {deviceOffer, deviceSelection, deviceFinished, devicePrimary};

static int writePayload(void) {
    const char* p    = payload;
    size_t      left = strlen(payload);
    while (left > 0) {
        const ssize_t n = write(send_fd, p, left);
        if (n > 0) {
            p += n;
            left -= (size_t)n;
            continue;
        }
        if (n < 0 && errno == EINTR)
            continue;
        break;
    }
    close(send_fd);
    send_fd = -1;
    return 1;
}

int main(int argc, char** argv) {
    if (argc != 3) {
        fprintf(stderr, "usage: cliphold DELAY_MS TEXT\n");
        return 2;
    }
    delay_ms = atoi(argv[1]);
    payload  = argv[2];
    signal(SIGPIPE, SIG_IGN);

    struct wl_display* display = wl_display_connect(NULL);
    if (!display)
        return 3;
    struct wl_registry* registry = wl_display_get_registry(display);
    wl_registry_add_listener(registry, &registryListener, NULL);
    if (wl_display_roundtrip(display) < 0 || !manager || !seat)
        return 4;

    struct zwlr_data_control_device_v1* device = zwlr_data_control_manager_v1_get_data_device(manager, seat);
    struct zwlr_data_control_source_v1* source = zwlr_data_control_manager_v1_create_data_source(manager);
    zwlr_data_control_device_v1_add_listener(device, &deviceListener, NULL);
    zwlr_data_control_source_v1_add_listener(source, &sourceListener, NULL);
    zwlr_data_control_source_v1_offer(source, "text/plain;charset=utf-8");
    zwlr_data_control_device_v1_set_selection(device, source);
    if (wl_display_roundtrip(display) < 0)
        return 5;
    puts("READY");
    fflush(stdout);

    const int displayFd = wl_display_get_fd(display);
    int       wrote     = 0;
    while (!cancelled && !wrote) {
        if (send_fd >= 0 && delay_ms >= 0 && monotonicMs() >= send_at) {
            wrote = writePayload();
            break;
        }
        struct pollfd pfd = {.fd = displayFd, .events = POLLIN};
        int           timeout = -1;
        if (send_fd >= 0 && delay_ms >= 0) {
            const int64_t left = send_at - monotonicMs();
            timeout            = left <= 0 ? 0 : left > INT32_MAX ? INT32_MAX : (int)left;
        }
        const int ready = poll(&pfd, 1, timeout);
        if (ready < 0 && errno == EINTR)
            continue;
        if (ready < 0 || (ready > 0 && (pfd.revents & (POLLERR | POLLHUP))))
            break;
        if (ready > 0 && (pfd.revents & POLLIN) && wl_display_dispatch(display) < 0)
            break;
        if (ready == 0)
            continue;
        wl_display_flush(display);
    }

    if (send_fd >= 0)
        close(send_fd);
    zwlr_data_control_source_v1_destroy(source);
    zwlr_data_control_device_v1_destroy(device);
    zwlr_data_control_manager_v1_destroy(manager);
    wl_seat_destroy(seat);
    wl_registry_destroy(registry);
    wl_display_disconnect(display);
    return 0;
}

File: devtools/src/bin/cliphold.rs
//! cliphold — nested-gate clipboard source that deliberately delays transfer.
//! It uses wlr-data-control so no surface or keyboard focus is needed.
//!
//!   `cliphold DELAY_MS TEXT`
//!
//! A negative delay holds the requested fd until the process or compositor
//! exits. READY is printed after the selection belongs to this client; SEND
//! is printed when a consumer requests it.
//!
//! Rust port of the C fixture (awesome-rust plan, Track D).

use std::os::fd::OwnedFd;
use std::time::{Duration, Instant};

use std::io::Write;

use rustix::event::Timespec;
use rustix::event::{PollFd, PollFlags, poll};
use rustix::io::write;
use wayland_client::backend::WaylandError;
use wayland_client::protocol::{wl_registry, wl_seat};
use wayland_client::{Connection, Dispatch, EventQueue, QueueHandle, event_created_child};
use wayland_protocols_wlr::data_control::v1::client::{
    zwlr_data_control_device_v1::{Event as DeviceEvent, ZwlrDataControlDeviceV1},
    zwlr_data_control_manager_v1::{Event as ManagerEvent, ZwlrDataControlManagerV1},
    zwlr_data_control_offer_v1::{Event as OfferEvent, ZwlrDataControlOfferV1},
    zwlr_data_control_source_v1::{Event as SourceEvent, ZwlrDataControlSourceV1},
};

// queue state: one data source, one pending transfer
struct State {
    manager: Option<ZwlrDataControlManagerV1>,
    seat: Option<wl_seat::WlSeat>,
    device: Option<ZwlrDataControlDeviceV1>,
    source: Option<ZwlrDataControlSourceV1>,
    delay_ms: i64,
    send_fd: Option<OwnedFd>,
    send_at: Option<Instant>,
    cancelled: bool,
}

impl State {
    fn delay(&self) -> Option<Duration> {
        (self.delay_ms >= 0)
            .then(|| Duration::from_millis(u64::try_from(self.delay_ms).unwrap_or(u64::MAX)))
    }

    const fn new(delay_ms: i64) -> Self {
        Self {
            manager: None,
            seat: None,
            device: None,
            source: None,
            delay_ms,
            send_fd: None,
            send_at: None,
            cancelled: false,
        }
    }
}

impl Dispatch<wl_registry::WlRegistry, ()> for State {
    fn event(
        state: &mut State,
        registry: &wl_registry::WlRegistry,
        event: wl_registry::Event,
        _data: &(),
        _conn: &Connection,
        qh: &QueueHandle<Self>,
    ) {
        if let wl_registry::Event::Global {
            name,
            interface,
            version,
        } = event
        {
            match interface.as_str() {
                "zwlr_data_control_manager_v1" if state.manager.is_none() => {
                    state.manager = Some(registry.bind::<ZwlrDataControlManagerV1, _, _>(
                        name,
                        version.min(2),
                        qh,
                        (),
                    ));
                }
                "wl_seat" if state.seat.is_none() => {
                    state.seat =
                        Some(registry.bind::<wl_seat::WlSeat, _, _>(name, version.min(7), qh, ()));
                }
                _ => {}
            }
        }
    }
}

impl Dispatch<ZwlrDataControlManagerV1, ()> for State {
    fn event(
        _state: &mut State,
        _proxy: &ZwlrDataControlManagerV1,
        _event: ManagerEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
    }
}

impl Dispatch<ZwlrDataControlDeviceV1, ()> for State {
    fn event(
        state: &mut State,
        _proxy: &ZwlrDataControlDeviceV1,
        event: DeviceEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
        match event {
            DeviceEvent::DataOffer { id: _offer } => {
                // an offer we never use; nothing to answer
            }
            // a foreign selection: C destroys the advertised offer and does
            // NOT cancel here — cancellation arrives via the source's
            // cancelled event (setting cancelled early changes the timing the
            // stale-transfer battery asserts on)
            DeviceEvent::Selection { id } | DeviceEvent::PrimarySelection { id } => {
                if let Some(offer) = id {
                    offer.destroy();
                }
            }
            DeviceEvent::Finished => state.cancelled = true,
            _ => {}
        }
    }
    // the data_offer event (opcode 0) creates a zwlr_data_control_offer_v1
    // child: without this specialization the queue panics on the first
    // foreign selection (the C fixture added an offer listener implicitly)
    event_created_child!(State, ZwlrDataControlDeviceV1, [
        0 => (ZwlrDataControlOfferV1, ()),
    ]);
}

impl Dispatch<ZwlrDataControlSourceV1, ()> for State {
    fn event(
        state: &mut State,
        _proxy: &ZwlrDataControlSourceV1,
        event: SourceEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
        match event {
            SourceEvent::Send { mime_type: _, fd } => {
                if let Some(old) = state.send_fd.replace(fd) {
                    drop(old);
                }
                // C: send_at = now + delay when delay >= 0, now + 0 otherwise
                // (the negative case never fires the write — see has_delay)
                state.send_at = Some(Instant::now() + (state.delay()).unwrap_or(Duration::ZERO));
                println!("SEND");
                let mut out = std::io::stdout();
                let _ = out.flush();
            }
            SourceEvent::Cancelled => state.cancelled = true,
            _ => {}
        }
    }
}

impl Dispatch<ZwlrDataControlOfferV1, ()> for State {
    fn event(
        _state: &mut State,
        _proxy: &ZwlrDataControlOfferV1,
        _event: OfferEvent,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
    }
}

impl Dispatch<wl_seat::WlSeat, ()> for State {
    fn event(
        _state: &mut State,
        _proxy: &wl_seat::WlSeat,
        _event: wl_seat::Event,
        _data: &(),
        _conn: &Connection,
        _qh: &QueueHandle<Self>,
    ) {
    }
}

// full write of the payload into the transfer fd (EINTR retried, a dead
// consumer ends the write like the C fixture's SIGPIPE ignore)
fn write_payload(fd: &OwnedFd, payload: &[u8]) {
    let mut offset = 0;
    while offset < payload.len() {
        match write(fd, &payload[offset..]) {
            Ok(n) => offset += n,
            Err(rustix::io::Errno::INTR) => {}
            Err(_) => break,
        }
    }
}

fn main() {
    let mut args = std::env::args().skip(1);
    let Some(delay_arg) = args.next() else {
        eprintln!("usage: cliphold DELAY_MS TEXT");
        std::process::exit(2);
    };
    let delay_ms: i64 = delay_arg.parse().unwrap_or_else(|_| {
        eprintln!("usage: cliphold DELAY_MS TEXT");
        std::process::exit(2);
    });
    let Some(payload) = args.next() else {
        eprintln!("usage: cliphold DELAY_MS TEXT");
        std::process::exit(2);
    };

    let Ok(conn) = Connection::connect_to_env() else {
        std::process::exit(3);
    };
    let mut queue: EventQueue<State> = conn.new_event_queue();
    let qh = queue.handle();
    conn.display().get_registry(&qh, ());

    let mut state = State::new(delay_ms);
    // wait for the globals
    loop {
        if queue.blocking_dispatch(&mut state).is_err() {
            std::process::exit(4);
        }
        if state.manager.is_some() && state.seat.is_some() {
            break;
        }
    }
    let (Some(manager), Some(seat)) = (state.manager.clone(), state.seat.clone()) else {
        std::process::exit(4);
    };

    let device = manager.get_data_device(&seat, &qh, ());
    let source = manager.create_data_source(&qh, ());
    state.device = Some(device);
    source.offer("text/plain;charset=utf-8".to_string());
    state.device.as_ref().unwrap().set_selection(Some(&source));
    state.source = Some(source);
    if queue.roundtrip(&mut state).is_err() {
        std::process::exit(5);
    }
    println!("READY");
    let mut out = std::io::stdout();
    let _ = out.flush();

    run_transfer(&mut queue, &mut state, delay_ms >= 0, payload.as_bytes());
    teardown(&mut state);
    let _ = conn.flush();
}

// a negative delay holds the fd until the process or compositor exits
fn run_transfer(
    queue: &mut EventQueue<State>,
    state: &mut State,
    has_delay: bool,
    payload_bytes: &[u8],
) {
    while !state.cancelled {
        // the delayed write fires when its deadline passes
        if let (Some(fd), Some(due)) = (&state.send_fd, state.send_at)
            && has_delay
            && Instant::now() >= due
        {
            write_payload(fd, payload_bytes);
            break;
        }
        let timeout = state
            .send_at
            .filter(|_| has_delay)
            .map(|due| due.saturating_duration_since(Instant::now()));
        let Some(guard) = queue.prepare_read() else {
            let _ = queue.dispatch_pending(state);
            continue;
        };
        let timeout_ts = timeout.map(|t| Timespec {
            tv_sec: i64::try_from(t.as_secs()).unwrap_or(i64::MAX),
            tv_nsec: i64::from(t.subsec_nanos()),
        });
        let conn_fd = guard.connection_fd();
        let mut fds = [PollFd::new(
            &conn_fd,
            PollFlags::IN | PollFlags::HUP | PollFlags::ERR,
        )];
        #[allow(clippy::match_same_arms)]
        match poll(&mut fds, timeout_ts.as_ref()) {
            Ok(0) => {}
            Ok(_) => {
                let ready = fds[0].revents();
                if ready.intersects(PollFlags::HUP | PollFlags::ERR)
                    && !ready.intersects(PollFlags::IN)
                {
                    break;
                }
                if let Err(err) = guard.read() {
                    let would_block = matches!(&err, WaylandError::Io(io) if io.kind() == std::io::ErrorKind::WouldBlock);
                    if !would_block {
                        eprintln!("cliphold: read: {err}");
                        break;
                    }
                }
                let _ = queue.flush();
            }
            Err(rustix::io::Errno::INTR) => {}
            Err(err) => {
                eprintln!("cliphold: poll: {err}");
                break;
            }
        }
        if let Err(err) = queue.dispatch_pending(state) {
            eprintln!("cliphold: dispatch: {err}");
            break;
        }
    }
}

fn teardown(state: &mut State) {
    state.send_fd.take();
    let _ = state.source.take().map(|source| source.destroy());
    let _ = state.device.take().map(|device| device.destroy());
    let _ = state.manager.take().map(|manager| manager.destroy());
    // wl_seat has no destroy request; dropping the proxy suffices
    state.seat.take();
}
